### PR TITLE
API Review Feedback

### PR DESCRIPTION
**Pull request by martintmk**
### Details on the issue fix or feature implementation

- Improved XML docs for arguments.
- Drop `OnHedgingArguments.HasOutcome` property
- Introduce `FallbackActionArguments` (instead of using `FallbackPredicateArguments` in `FallbackAction` delegate)
- Rename `RetryDelayArguments` to `RetryDelayGeneratorArguments` 
- Rename `HedgingDelayArguments` to `HedgingDelayGeneratorArguments` 
- Allow registering multiple listeners in `TelemetryOptions`

Contributes to #1507 

### Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

File: src/Polly.Core/CircuitBreaker/CircuitBreakerPredicateArguments.cs [hunk 1]
```diff
@@ -23,12 +23,12 @@ public CircuitBreakerPredicateArguments(ResilienceContext context, Outcome<TResu
     }
 
     /// <summary>
-    /// Gets the outcome of the resilience operation or event.
+    /// Gets the outcome of user-callback.
```
Comment (martincostello) [inline src/Polly.Core/CircuitBreaker/CircuitBreakerPredicateArguments.cs:26]:
```suggestion
    /// Gets the outcome of the user-specified callback.
```

File: src/Polly.Core/Fallback/FallbackPredicateArguments.cs [hunk 2]
```diff
@@ -23,12 +23,12 @@ public FallbackPredicateArguments(ResilienceContext context, Outcome<TResult> ou
     }
 
     /// <summary>
-    /// Gets the outcome of the resilience operation or event.
+    /// Gets the outcome of user-callback.
```
Comment (martincostello) [inline src/Polly.Core/Fallback/FallbackPredicateArguments.cs:26]:
```suggestion
    /// Gets the outcome of the user-specified callback.
```

File: src/Polly.Core/Hedging/HedgingPredicateArguments.cs [hunk 3]
```diff
@@ -23,12 +23,12 @@ public HedgingPredicateArguments(ResilienceContext context, Outcome<TResult> out
     }
 
     /// <summary>
-    /// Gets the outcome of the resilience operation or event.
+    /// Gets the outcome of user-callback.
```
Comment (martincostello) [inline src/Polly.Core/Hedging/HedgingPredicateArguments.cs:26]:
```suggestion
    /// Gets the outcome of the user-specified callback.
```

File: src/Polly.Core/Retry/RetryPredicateArguments.cs [hunk 4]
```diff
@@ -25,12 +25,12 @@ public RetryPredicateArguments(ResilienceContext context, Outcome<TResult> outco
     }
 
     /// <summary>
-    /// Gets the outcome of the resilience operation or event.
+    /// Gets the outcome of user-callback.
```
Comment (martincostello) [inline src/Polly.Core/Retry/RetryPredicateArguments.cs:28]:
```suggestion
    /// Gets the outcome of the user-specified callback.
```

File: src/Polly.Extensions/Telemetry/TelemetryListenerImpl.cs [hunk 5]
```diff
@@ -43,7 +43,13 @@ public TelemetryListenerImpl(TelemetryOptions options)
 
     public override void Write<TResult, TArgs>(in TelemetryEventArguments<TResult, TArgs> args)
     {
-        _listener?.Write(in args);
+        if (_listeners.Count > 0)
+        {
+            foreach (var listener in _listeners)
+            {
+                listener.Write(in args);
+            }
```
Comment (martincostello) [inline src/Polly.Extensions/Telemetry/TelemetryListenerImpl.cs:51]: As this is a `List<T>`, would using its own `ForEach()` method be more performant than using a `foreach`?

```suggestion
            _listeners.ForEach((listener) => listener.Write(in args));
```

File: test/Polly.Core.Tests/Hedging/HedgingResiliencePipelineBuilderExtensionsTests.cs [hunk 6]
```diff
@@ -88,9 +88,9 @@ public async Task AddHedging_IntegrationTest()
                 },
                 OnHedging = args =>
                 {
-                    if (args.HasOutcome)
+                    if (args.Outcome is not null)
                     {
-                        results.Enqueue(args.Outcome.Result!.ToString()!);
+                        results.Enqueue(args.Outcome!.Value.Result!.ToString()!);
```
Comment (martincostello) [inline test/Polly.Core.Tests/Hedging/HedgingResiliencePipelineBuilderExtensionsTests.cs:93]: What about:

```csharp
if (args.Outcome is { } outcome)
{
    results.Enqueue(outcome.Value.Result!.ToString()!);
}
```

Comment (martintmk) [reply test/Polly.Core.Tests/Hedging/HedgingResiliencePipelineBuilderExtensionsTests.cs:93]: indeed, looks much nicer.

File: src/Polly.Extensions/Telemetry/TelemetryListenerImpl.cs (hunk 5 above)
Comment (martintmk) [reply src/Polly.Extensions/Telemetry/TelemetryListenerImpl.cs:51]: It would closure allocation because of `args`.

**Review by martincostello — APPROVED** [review]

**Pull request merged**